### DISCUSSION TITLE
Qdrant memory and specs requirement for billions of vectors

### DISCUSSION THREAD

**Discussion by TheRabidWolverine** [qdrant/qdrant#2607] [Q&A]
I came across [this](https://qdrant.tech/articles/memory-consumption/#) article describing Qdrant's RAM and other configuration benchmarking experiments. Looked pretty cool. I have some questions before we start using it for our project, since we all are fairly new to this, and these answers may help us evaluate our choices and configurations.

1. The article mentions some 135 MB of RAM for 1 million vectors with good performance under SSD. It does not quantify "good". How good is "good" here? Is that production level, like the kind one can see on Google or Bing image search result page, for example, when one searches for suppose "tiger cubs playing with each other" and immediately gets hundreds of matching images?

2. The above example states 128 dimensions. Can we assume that for 512 (128 * 4) dimensions, an equivalent performance will need 135 * 4 = 540 MB of RAM as an approximation? Will the calculation be linear like that? Or will there be checkpoints or mountpoints in the number of vectors increasing (or their dimensions) where the requirement may spike up or plateau down?

3. If 135 MB is indeed for 1 million vectors, then extrapolating that 1 billion vectors, it becomes 135 GB. That's just about 1 single EC2 node with 150 GB RAM. Will 135 GB RAM, coupled with sufficiently high amount of SSD, be enough to give production level vector comparison queries at runtime for 1 billion vectors? Milvus, for example, suggests [here](https://milvus.io/tools/sizing/) an entire arsenal of resources required to cater to 1 billion vectors, whereas, here, just one machine will suffice? Am I missing something?

4. [This](https://zilliz.com/comparison/qdrant-vs-milvus) page, although authored by Ziliz, rendering it prone to bias, suggests that Milvus can dynamically scale and shard up, and cater to billion plus vectors, while Qdrant cannot. Whereas your [page](https://qdrant.tech/benchmarks/) suggests that Qdrant is better at everything. Which one is true?

**Comment by generall** [comment]
>  It does not quantify "good".

This article only speaks about speed of retrieval, it does not make any statement about quality of the search, as it mostly depends on the model you are using. Also, there are no mentions of "good" in it.

Regarding speed - there is a table with estimations in this section: https://qdrant.tech/articles/memory-consumption/#how-to-speed-up-the-search

>  Will the calculation be linear like that? 

Yes, for vector dimensionality it is close to linear

> If 135 MB is indeed for 1 million vectors, then extrapolating that 1 billion vectors, it becomes 135 GB. That's just about 1 single EC2 node with 150 GB RAM. Will 135 GB RAM, coupled with sufficiently high amount of SSD, be enough to give production level vector comparison queries at runtime for 1 billion vectors?


Yes, one machine of that size is enough to serve 1 billion 128d vectors, we checked it in our experiments. Of course, if you have additional requirement, like constant insertion of new vectors, or high RPS, or millisecond latency - it might require more resources.
So the one machine of 150Gb is minimal requirement.

> suggests that Milvus can dynamically scale and shard up, and cater to billion plus vectors, while Qdrant cannot

Qdrant doesn't yet have dymanic re-sharding, meaning that you can't yet change the number of shards in the existing collection.
However I am not sure how much of a limitation it is in practice,  considering that both vertical and horizontal scaling is possible within the pre-configured range, e.g. if you pre-configure 4 shards, you can scale from 1x2Gb machine into 4x64Gb (x128 times) with no downtime

**Reply by TheRabidWolverine** [reply]
Thanks. That was very helpful. Some questions:

1. You mentioned sharding. Does that span machines? Say in our above example, we have one 150 GB RAM machine, and we are storing 1 billion vectors. It occupies 135 GB of RAM approximately, and 15 GB is left as spare. As more images keep pouring in, the size keeps increasing. When it reaches 149 GB, we decide to commission a new machine. Will that also need to be configured with say 4 shards, like the earlier machine? 

2. You mentioned dimensionality is linear. Does that also mean that 135 GB is able to accommodate 1 billion 128D vectors, around 135 * 4 = 540 GB will be required to accommodate 1 billion 512D vectors as well? Probably yes, just checking!

3. You also mentioned quality of search. Imagine some established model like OpenAI'c CLIP is being deployed, that guarantees good quality, so will that, combined with high speed of retrieval as you said, will be able to power a good image searching backend?

**Reply by generall** [reply]
> Will that also need to be configured with say 4 shards, like the earlier machine?

At that scale you might need even more shards (for faster indexing). In our experiment we were using 8 shards on one machine.
So whenever your 1 machine is full, you would need to spawn another one and move half of the shards there (there is an API for that)

> 540 GB will be required to accommodate 1 billion 512D vectors as well? Probably yes, just checking!

Yes, but keep in mind that those estimations assuming that larger part of vectors is going to be stored on disk, and disk is expected to be quite fast. At least 100k IOPS.

>  Imagine some established model like OpenAI'c CLIP is being deployed, that guarantees good quality, so will that, combined with high speed of retrieval as you said, will be able to power a good image searching backend?

Not sure if CLIP actually guarantees anything, but if you are satisfied with that quality, then yes.

**Reply by TheRabidWolverine** [reply]
Thanks, when large part of the vectors are stored on disk, even with SSD, can the performance be good enough to be production like, the kinds you get on Google or Bing image search pages? Will 180 GB of RAM, in our scenario, where 135 is the approximate requirement, make things faster by being able to offload a large part of the supposed-to-be-on-the-disk vectors to the remaining 45 GB of spare memory? The average SSD one gets on say EC2 or Azure, do they serve the 100k IOPS perf you mentioned? Will that, clubbed with the 135 GB of RAM be enough to yield production level perf?

With CLIP, I mean, yes, at the end of the day it boils down to the quality of our model and training, which we need to do separately on high end GPUs like A100s, which we will do and then deploy on these 150 GB machines for querying. So, yes, the question was more for the speed of similarity search. :)

**Reply by generall** [reply]
> good enough to be production like, the kinds you get on Google or Bing image search pages?

Google have million request per second on it's servers. One machine with disk-based deployment can do maybe 100 RPS best case scenario.

> offload a large part of the supposed-to-be-on-the-disk vectors to the remaining 45 GB of spare memory?

There will be no spare memory. All available memory will be used for disk cache.l

>  The average SSD one gets on say EC2 or Azure, do they serve the 100k IOPS perf you mentioned?

Only if it is local disk. Network mounted volumes won't be enough

**Reply by TheRabidWolverine** [reply]
Makes sense, thanks. To increase RPS, perhaps being able to accommodate 100s of vectors per second, what needs to be increased - RAM, I guess? Or more nodes with shards, and thus the load will be automatically more distributed and hence each node will be able to accommodate more RPS (runtime queries) while also accommodating fairly high number of vectors being inserted per second?

Secondly, we figured out from the 135 MB figure than 1 billion vectors will need around 135 GB of RAM. So all the vectors are fitting in the memory already, right? Why should there be no spare memory if the total available memory is 180 GB? What will the remaining 45 GB memory be used by Qdrant for?

**Reply by generall** [reply]
> Secondly, we figured out from the 135 MB figure than 1 billion vectors will need around 135 GB of RAM. So all the vectors are fitting in the memory already, right? 

No, in this scenario almost no vectors are in RAM. To put 1B of 128d vectors in ram you would need about 500Gb

**Reply by TheRabidWolverine** [reply]
Umm, with almost no vector in RAM, will the performance be production level? And Milvus asks for much more memory - https://milvus.io/tools/sizing/, is it because, unlike Qdrant, it assumes that it will keep majority of the data in memory? And the Qdrant blogpost also mentions that is has better perfs than Milvus in almost all aspects, so, Qdrant, with a majority of its data on disk, performs better than Milvus, which takes much more memory, and is likely to have a substantial part of its data in memory? How come?

**Reply by generall** [reply]
> Umm, with almost no vector in RAM, will the performance be production level?


Again, that depends on your definition of `production level`. In particular - required latency and number of requests per second. 
It also depends on disk performance you are going to have.
